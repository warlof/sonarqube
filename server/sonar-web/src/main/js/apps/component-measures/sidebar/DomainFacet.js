/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import FacetBox from '../../../components/facet/FacetBox';
import FacetHeader from '../../../components/facet/FacetHeader';
import FacetItem from '../../../components/facet/FacetItem';
import FacetItemsList from '../../../components/facet/FacetItemsList';
import FacetMeasureValue from './FacetMeasureValue';
import IssueTypeIcon from '../../../components/ui/IssueTypeIcon';
import Tooltip from '../../../components/controls/Tooltip';
import { filterMeasures, sortMeasures } from '../utils';
import { getLocalizedMetricDomain, getLocalizedMetricName } from '../../../helpers/l10n';
import type { MeasureEnhanced } from '../../../components/measure/types';

type Props = {|
  onChange: (metric: string) => void,
  onToggle: (property: string) => void,
  open: boolean,
  domain: { name: string, measures: Array<MeasureEnhanced> },
  selected: string
|};

export default class DomainFacet extends React.PureComponent {
  props: Props;

  handleHeaderClick = () => this.props.onToggle(this.props.domain.name);

  render() {
    const { domain, selected } = this.props;
    const measures = sortMeasures(domain.name, filterMeasures(domain.measures));
    return (
      <FacetBox>
        <FacetHeader
          name={getLocalizedMetricDomain(domain.name)}
          onClick={this.handleHeaderClick}
          open={this.props.open}
          values={measures.find(measure => measure.metric.key === selected) ? 1 : 0}
        />

        {this.props.open &&
          <FacetItemsList>
            {measures.map(measure =>
              <FacetItem
                active={measure.metric.key === selected}
                disabled={false}
                key={measure.metric.key}
                name={
                  <Tooltip overlay={getLocalizedMetricName(measure.metric)} mouseEnterDelay={1}>
                    <span id={`measure-${measure.metric.key}-name`}>
                      <IssueTypeIcon query={measure.metric.key} className="little-spacer-right" />
                      {getLocalizedMetricName(measure.metric)}
                    </span>
                  </Tooltip>
                }
                onClick={this.props.onChange}
                stat={<FacetMeasureValue measure={measure} />}
                value={measure.metric.key}
              />
            )}
          </FacetItemsList>}
      </FacetBox>
    );
  }
}
