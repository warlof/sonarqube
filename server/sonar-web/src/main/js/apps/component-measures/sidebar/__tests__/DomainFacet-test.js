/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import { shallow } from 'enzyme';
import DomainFacet from '../DomainFacet';

const DOMAIN = {
  name: 'Reliability',
  measures: [
    {
      metric: {
        key: 'bugs',
        type: 'INT',
        name: 'Bugs',
        domain: 'Reliability'
      },
      value: '5',
      periods: [{ index: 1, value: '5' }],
      leak: '5'
    },
    {
      metric: {
        key: 'new_bugs',
        type: 'INT',
        name: 'New Bugs',
        domain: 'Reliability'
      },
      periods: [{ index: 1, value: '5' }],
      leak: '5'
    }
  ]
};

const PROPS = {
  onChange: () => {},
  onToggle: () => {},
  open: true,
  domain: DOMAIN,
  selected: 'foo'
};

it('should display facet item list', () => {
  expect(shallow(<DomainFacet {...PROPS} />)).toMatchSnapshot();
});

it('should display facet item list with bugs selected', () => {
  expect(shallow(<DomainFacet {...PROPS} selected="bugs" />)).toMatchSnapshot();
});

it('should render closed', () => {
  const wrapper = shallow(<DomainFacet {...PROPS} open={false} />);
  expect(wrapper.find('FacetItemsList')).toHaveLength(0);
});
