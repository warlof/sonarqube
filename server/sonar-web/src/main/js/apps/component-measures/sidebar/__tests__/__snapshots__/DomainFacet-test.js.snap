// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display facet item list 1`] = `
<FacetBox>
  <FacetHeader
    name="Reliability"
    onClick={[Function]}
    open={true}
    values={0}
  />
  <FacetItemsList>
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      name={
        <Tooltip
          mouseEnterDelay={1}
          overlay="Bugs"
          placement="bottom"
        >
          <span
            id="measure-bugs-name"
          >
            <IssueTypeIcon
              className="little-spacer-right"
              query="bugs"
            />
            Bugs
          </span>
        </Tooltip>
      }
      onClick={[Function]}
      stat={
        <FacetMeasureValue
          measure={
            Object {
              "leak": "5",
              "metric": Object {
                "domain": "Reliability",
                "key": "bugs",
                "name": "Bugs",
                "type": "INT",
              },
              "periods": Array [
                Object {
                  "index": 1,
                  "value": "5",
                },
              ],
              "value": "5",
            }
          }
        />
      }
      value="bugs"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      name={
        <Tooltip
          mouseEnterDelay={1}
          overlay="New Bugs"
          placement="bottom"
        >
          <span
            id="measure-new_bugs-name"
          >
            <IssueTypeIcon
              className="little-spacer-right"
              query="new_bugs"
            />
            New Bugs
          </span>
        </Tooltip>
      }
      onClick={[Function]}
      stat={
        <FacetMeasureValue
          measure={
            Object {
              "leak": "5",
              "metric": Object {
                "domain": "Reliability",
                "key": "new_bugs",
                "name": "New Bugs",
                "type": "INT",
              },
              "periods": Array [
                Object {
                  "index": 1,
                  "value": "5",
                },
              ],
            }
          }
        />
      }
      value="new_bugs"
    />
  </FacetItemsList>
</FacetBox>
`;

exports[`should display facet item list with bugs selected 1`] = `
<FacetBox>
  <FacetHeader
    name="Reliability"
    onClick={[Function]}
    open={true}
    values={1}
  />
  <FacetItemsList>
    <FacetItem
      active={true}
      disabled={false}
      halfWidth={false}
      name={
        <Tooltip
          mouseEnterDelay={1}
          overlay="Bugs"
          placement="bottom"
        >
          <span
            id="measure-bugs-name"
          >
            <IssueTypeIcon
              className="little-spacer-right"
              query="bugs"
            />
            Bugs
          </span>
        </Tooltip>
      }
      onClick={[Function]}
      stat={
        <FacetMeasureValue
          measure={
            Object {
              "leak": "5",
              "metric": Object {
                "domain": "Reliability",
                "key": "bugs",
                "name": "Bugs",
                "type": "INT",
              },
              "periods": Array [
                Object {
                  "index": 1,
                  "value": "5",
                },
              ],
              "value": "5",
            }
          }
        />
      }
      value="bugs"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      name={
        <Tooltip
          mouseEnterDelay={1}
          overlay="New Bugs"
          placement="bottom"
        >
          <span
            id="measure-new_bugs-name"
          >
            <IssueTypeIcon
              className="little-spacer-right"
              query="new_bugs"
            />
            New Bugs
          </span>
        </Tooltip>
      }
      onClick={[Function]}
      stat={
        <FacetMeasureValue
          measure={
            Object {
              "leak": "5",
              "metric": Object {
                "domain": "Reliability",
                "key": "new_bugs",
                "name": "New Bugs",
                "type": "INT",
              },
              "periods": Array [
                Object {
                  "index": 1,
                  "value": "5",
                },
              ],
            }
          }
        />
      }
      value="new_bugs"
    />
  </FacetItemsList>
</FacetBox>
`;
